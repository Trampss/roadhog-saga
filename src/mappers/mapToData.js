export default fallback => function* (raw) {
  if (!raw.ok && fallback) return fallback
  if (raw.status !== 204) return yield raw.json()
  return undefined
}
