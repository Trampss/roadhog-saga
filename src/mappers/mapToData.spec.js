/* eslint-env jest */
import tester from 'trampss-redux-saga-tester'
import mapToData from './mapToData'

describe('mapToData', () => {
  const fallback = { fallback: true }
  const error = 'PARSING ERROR'
  const onError = () => { throw new Error(error) }

  describe('with json content-type', () => {
    const jsonRaw = { headers: new Map([['content-type', 'application/json']]), json: () => ({ data: 'JSON' }), text: () => 'TEXT' }

    it('should return undefined -code 204 no content-', () => {
      const raw = { ...jsonRaw, ok: true, status: 204 }
      const test = tester(mapToData(fallback))(raw)(/* no mock */)
      expect(test).toMatchSnapshot()
    })

    it('should return result -ok-', () => {
      const raw = { ...jsonRaw, ok: true }
      const test = tester(mapToData(fallback))(raw)(/* no mock */)
      expect(test).toMatchSnapshot()
    })

    it('should return fallback -ko-', () => {
      const raw = { ...jsonRaw, ok: false }
      const test = tester(mapToData(fallback))(raw)(/* no mock */)
      expect(test).toMatchSnapshot()
    })

    it('should return result -ko-', () => {
      const raw = { ...jsonRaw, ok: false }
      const test = tester(mapToData(/* no fallback */))(raw)(/* no mock */)
      expect(test).toMatchSnapshot()
    })

    it('should return text with json parse on error', () => {
      const raw = { ...jsonRaw, ok: true, json: onError }
      const test = tester(mapToData(fallback))(raw)(/* no mock */)
      expect(test).toMatchSnapshot()
    })

    it('should throw error when json parse and text parse on error', () => {
      const raw = { ...jsonRaw, ok: true, json: onError, text: onError }
      const test = () => tester(mapToData(fallback))(raw)(/* no mock */)
      expect(test).toThrowErrorMatchingSnapshot()
    })
  })

  describe('with plain/text content-type', () => {
    const textRaw = { headers: new Map([['content-type', 'text/plain']]), json: onError, text: () => 'TEXT' }

    it('should return result -ok-', () => {
      const raw = { ...textRaw, ok: true }
      const test = tester(mapToData(fallback))(raw)(/* no mock */)
      expect(test).toMatchSnapshot()
    })

    it('should return fallback -ko-', () => {
      const raw = { ...textRaw, ok: false }
      const test = tester(mapToData(fallback))(raw)(/* no mock */)
      expect(test).toMatchSnapshot()
    })

    it('should return result -ko-', () => {
      const raw = { ...textRaw, ok: false }
      const test = tester(mapToData(/* no fallback */))(raw)(/* no mock */)
      expect(test).toMatchSnapshot()
    })

    it('should throw error when text parse on error', () => {
      const raw = { ...textRaw, ok: true, text: onError }
      const test = () => tester(mapToData(fallback))(raw)(/* no mock */)
      expect(test).toThrowErrorMatchingSnapshot()
    })
  })

  describe('with no content-type', () => {
    it('should return text when content-type is undefined', () => {
      const raw = { ok: true, headers: new Map(), text: () => 'TEXT' }
      const test = tester(mapToData(fallback))(raw)(/* no mock */)
      expect(test).toMatchSnapshot()
    })
  })
})
