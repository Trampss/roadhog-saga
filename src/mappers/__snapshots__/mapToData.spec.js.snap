exports[`mapToData with json content-type should return fallback -ko- 1`] = `
Array [
  Object {
    "done": true,
    "value": Object {
      "fallback": true,
    },
  },
]
`;

exports[`mapToData with json content-type should return result -ko- 1`] = `
Array [
  Object {
    "done": false,
    "value": "TEXT",
  },
  Object {
    "done": true,
    "value": "TEXT",
  },
]
`;

exports[`mapToData with json content-type should return result -ok- 1`] = `
Array [
  Object {
    "done": false,
    "value": "TEXT",
  },
  Object {
    "done": true,
    "value": "TEXT",
  },
]
`;

exports[`mapToData with json content-type should return text with json parse on error 1`] = `
Array [
  Object {
    "done": false,
    "value": "TEXT",
  },
  Object {
    "done": true,
    "value": "TEXT",
  },
]
`;

exports[`mapToData with json content-type should return undefined -code 204 no content- 1`] = `
Array [
  Object {
    "done": true,
    "value": undefined,
  },
]
`;

exports[`mapToData with json content-type should throw error when json parse and text parse on error 1`] = `"PARSING ERROR"`;

exports[`mapToData with no content-type should return text when content-type is undefined 1`] = `
Array [
  Object {
    "done": false,
    "value": "TEXT",
  },
  Object {
    "done": true,
    "value": "TEXT",
  },
]
`;

exports[`mapToData with plain/text content-type should return fallback -ko- 1`] = `
Array [
  Object {
    "done": true,
    "value": Object {
      "fallback": true,
    },
  },
]
`;

exports[`mapToData with plain/text content-type should return result -ko- 1`] = `
Array [
  Object {
    "done": false,
    "value": "TEXT",
  },
  Object {
    "done": true,
    "value": "TEXT",
  },
]
`;

exports[`mapToData with plain/text content-type should return result -ok- 1`] = `
Array [
  Object {
    "done": false,
    "value": "TEXT",
  },
  Object {
    "done": true,
    "value": "TEXT",
  },
]
`;

exports[`mapToData with plain/text content-type should throw error when text parse on error 1`] = `"PARSING ERROR"`;
